docs/architecture/adr-051-double-signing-protection-with-tendermint-mode.md

```diff
@@ -0,0 +1,104 @@
+# ADR 51: Double Signing Protection with Tendermint Mode
+
+## Changelog
+* 27-11-2019: Initial draft
+
+## Context
+
+There exists two main purposes for this ADR.
+
+1. To introduce the simplest functional abstraction of Tendermint Mode
+    - Fullnode mode : fullnode mode does not have any capability to participate on consensus
+    - Validator mode : this mode is exactly same as existing state machine behavior. sync without voting on consensus, and participate consensus when fully synced
+    - Seed mode : lightweight seed mode only for maintain an address book, p2p like [TenderSeed](https://gitlab.com/polychainlabs/tenderseed)
+2. To provide a prevention method for double signing incidents mistakenly executed by validators
+    - Validators often mistakenly run duplicated validators to cause double-signing incident
+    - This proposed feature is to reduce the risk of mistaken double-signing incident by checking recent N blocks before voting begins
+    - When we think of such serious impact on double-signing incident, it is very reasonable to have multiple protection algorithm built in node daemon
+
+## Decision
+
+We would like to suggest a simple Tendermint mode abstraction with a double signing prevention method.
+
+### 1 ) Tendermint Mode
+
+- Which reactor, component to include for each node
+    - fullnode *(default)*
+        - switch, transport
+        - all reactors
+        - mempool, state
+        - rpc
+        - *~~no privValidator(priv_validator_key.json, priv_validator_state.json)~~*
+    - validator
+        - switch, transport
+        - all reactors
+        - mempool, state
+        - rpc
+        - with privValidator(priv_validator_key.json, priv_validator_state.json)
+    - seed
+        - switch, transport
+        - only pexReactor
+- Configuration, cli command
+    - We would like to suggest by introducing `mode` parameter in `config.toml` and cli
+    - `mode = "{{ .BaseConfig.Mode }}"` in `config.toml`
+    - `tendermint node --mode validator`  in cli
+    - fullnode | validator | seed (default: "fullnode")
+- RPC modification
+    - `host:26657/status`
+        - return empty `validator_info` when fullnode mode
+        - add return `mode` of this node
+    - no rpc server in seed mode
+- Where to modify in codebase
+    - Add  switch for `config.Mode` on `node/node.go:DefaultNewNode`
+    - If `config.Mode==validator`, call default `NewNode` (current logic)
+    - If `config.Mode==fullnode`, call `NewNode` with `nil` `privValidator` (do not load or generation)
+        - Need to add exception routine for `nil` `privValidator` to related functions
+    - If `config.Mode==seed`, call `NewSeedNode` (seed version of `node/node.go:NewNode`)
```

Comment:
I think we should extract seed mode completely into its own package

Reply:
you mean it should be independent binary for seed mode, not flag option?

Reply:
no, I meant `Go package` ~ extract seed mode code from p2p/pex/pex_reactor.go into its own Go package.

Reply:
oh I see, that would be better

Reply:
was it decided how this will be handled? in adr-052 it says under one binary but here there is a conversation to have it separated.

Reply:
I think what melekes said means not a binary separation but a refactoring of the go package directory structure, right?

Reply:
ah whoops sorry misread this